Optimize filter condition with switch case

maswin opened the pull request:
Test plan - Unit tests + verifier run on internal queries

```
== RELEASE NOTES ==

General Changes
* Added a new Optimizer SimplifySwitchExpression which simplifies switch cases in filter condition
```

The DB visualization tools such as looker & DbVisualizer generates sub-optimal queries when a selected column has switch case and a filter is applied on top of it. For instance it generates queries such as:

`SELECT case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral' from students where (case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral') = 'good';`

But this query can be simplified into:

`SELECT case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral' from students where behavior_code = 1;`

Without this simplification, the query misses out a lot of other optimizations (i.e, ORC indexes) and becomes extremely slow.

Since these are generated queries based on the UI, the user doesn't have much control on them.

kaikalur commented:
An interesting way to do this would be to flip WHEN/THEN (by appropriately adding equals to the THEN) and calling RowExpresionInterpreter. Like take the example and create a new expression:

```
case 'good' = 'good' then behavior_code=1 when 'good'= 'bad' then behavior_code=2  when 'good' =  'neutral' then not(behavior_code=1  OR behavior_code=2 ) end
```

That should be more robust.

kaikalur commented:
@rongrong Check it out

kaikalur commented:
In fact, this can be generalized to any relational expression where one side is case - just push relation to the THEN parts, flip them and evaluate it. ELSE part is simply NOT of OR of all the new THEN parts.

maswin commented:
What about the case when 2 WHEN clauses match the THEN part
```
(CASE 
  WHEN column_val = 1 THEN 'a'   // Matching
  WHEN column_val = 2 THEN 'a'   // Matching
  WHEN column_val = 3 THEN 'b'
ELSE 'c') = 'a'
```
flipping this will make it:
```
CASE 
  WHEN 'a'='a' THEN column_val = 1
  WHEN 'a'='a' THEN column_val = 2
  WHEN 'a'='b' THEN column_val = 3
  WHEN 'a'='c' THEN !(column_val = 1 OR column_val = 2 OR column_val = 3)
ELSE false
```
Since in switch case when first condition is matched we use it and break of, RowExpressionInterpter will simplify this to:
`column_val = 1`
But the right simplification is
`column_val = 1 OR column_val = 2`

maswin commented:
It needs to be converted to a series of OR clauses:
```
(CASE 
  WHEN column_val = 1 THEN 'a'   // Matching
  WHEN column_val = 2 THEN 'a'   // Matching
  WHEN column_val = 3 THEN 'b'
ELSE 'c') = 'a'
```

above expression gets transformed to:

```
('a'='a' AND column_val = 1) OR 
('a'='a' AND column_val = 2) OR 
('b'='a' AND column_val = 3) OR 
('c'='a' AND !(column_val = 1 OR column_val = 2 OR column_val = 3))
```

RowExpressionInterpter can transform this to a simpler expression

`column_val = 1 OR column_val = 2`

This handles lot of other extreme cases too, 
i.e one of the THEN clause and ELSE clause matches
```
(CASE 
  WHEN column_val = 1 THEN 'a'   // Matching
  WHEN column_val = 2 THEN 'b' 
ELSE 'a'      // Matching
) = 'a' 
```
get simplified to
```
column_val != 2
```

kaikalur commented:
> It needs to be converted to a series of OR clauses:
> 
> ```
> (CASE 
>   WHEN column_val = 1 THEN 'a'   // Matching
>   WHEN column_val = 2 THEN 'a'   // Matching
>   WHEN column_val = 3 THEN 'b'
> ELSE 'c') = 'a'
> ```
> 
> above expression gets transformed to:
> 
> ```
> ('a'='a' AND column_val = 1) OR 
> ('a'='a' AND column_val = 2) OR 
> ('b'='a' AND column_val = 3) OR 
> ('c'='a' AND !(column_val = 1 OR column_val = 2 OR column_val = 3))
> ```
> 
> RowExpressionInterpter can transform this to a simpler expression
> 
> `column_val = 1 OR column_val = 2`
> 
> This handles lot of other extreme cases too, i.e one of the THEN clause and ELSE clause matches
> 
> ```
> (CASE 
>   WHEN column_val = 1 THEN 'a'   // Matching
>   WHEN column_val = 2 THEN 'b' 
> ELSE 'a'      // Matching
> ) = 'a' 
> ```
> 
> get simplified to
> 
> ```
> column_val != 2
> ```

Hmm, yeah I'm usually not a fan of OR expressions because they are harder to pushdown. If we are going to stick to doing it only for cases when *all* THEN parts are constant/literal, we can just build a map THEN -> WHEN[] and the generate an OR for each.

maswin commented:
> Hmm, yeah I'm usually not a fan of OR expressions because they are harder to pushdown. If we are going to stick to doing it only for cases when _all_ THEN parts are constant/literal, we can just build a map THEN -> WHEN[] and the generate an OR for each.

Yeah, it is difficult to pushdown, but atleast OR seems better to deal with than  CASE expressions for further Optimizations. Was assuming if not a non deterministic function is involved in the expression to do this optimization. Since conversion from CASE to OR doesn't change the final evaluation result of the predicate and scenarios like concat('ab'. 'cd') = 'abcd' will also be handled

kaikalur commented:
> > Hmm, yeah I'm usually not a fan of OR expressions because they are harder to pushdown. If we are going to stick to doing it only for cases when _all_ THEN parts are constant/literal, we can just build a map THEN -> WHEN[] and the generate an OR for each.
> 
> Yeah, it is difficult to pushdown, but atleast OR seems better to deal with than CASE expressions for further Optimizations. Was assuming if not a non deterministic function is involved in the expression to do this optimization. Since conversion from CASE to OR doesn't change the final evaluation result of the predicate and scenarios like concat('ab'. 'cd') = 'abcd' will also be handled

Actually CASE is sequential and OR is not so but I guess it won't matter here if we are checking for constants only. Issue with OR is the whole NULL handling. So if we can do CASE, we should. 

Also like I said see if you can generalize for all relational operators.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (functionResolution.isComparisonFunction(node.getFunctionHandle())) {
+                if (node.getArguments().size() == 2) {
```
Combine the two if's into AND for better readability

maswin replied:
Done.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+            return LogicalRowExpressions.or(andExpressions);
+        }
+
+        private static class SwitchExpression
```
Do you really need this class?

maswin replied:
Simplified without the class.

kaikalur reviewed (changes requested):
Also please add some explicit end-to-end tests in hive connector to see the impact of this optimization.

kaikalur commented:
In fact, we want to check either all of the THEN and ELSE are constant or the otherExpression is constant

kaikalur commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -810,6 +811,11 @@ public SystemSessionProperties(
+                        OPTIMIZE_CASE_EXPRESSION_PREDICATE,
+                        "Optimize case expression predicates",
+                        featuresConfig.isOptimizeCaseExpressionPredicate(),
+                        true),
```
And we normally start by disabling the feature by default

maswin replied:
Done.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        private RowExpression processCaseExpression(SwitchExpression switchExpression,
+                                                    RowExpression otherExpression,
```
Done.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+                                             RowExpression right,
+                                             String displayName,
+                                             FunctionHandle functionHandle)
+        {
```
Moved to LogicalRowExpressions.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+            {
+                RowExpression caseExpression = specialForm.getArguments().get(0);
+                boolean isCaseExpressionAvailable = !(caseExpression instanceof ConstantExpression &&
+                        ((ConstantExpression) caseExpression).getValue().equals(true));
```
Both SearchedCase and SimpleCase expression both have the same format in RowExpresion, if its a SimpleCase Expression, in RowExpresion in place of the case operand we would have ConstantExpression(true). So I am performing the check to avoid doing true = when condition. 
Similar to this: https://github.com/prestodb/presto/blob/123c2ab8487aa962344601895d98af6fec22ddb3/presto-main/src/main/java/com/facebook/presto/sql/gen/SwitchCodeGenerator.java#L83

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,223 @@
+                                                    FunctionHandle functionHandle,
+                                                    String displayName,
+                                                    boolean switchExpressionOnRight)
+        {
```
We have scenarios such as 
```
(CASE
    WHEN (replace(column_name, 'prefix_string')) = 'dept1' THEN 'V1'
    WHEN (replace(column_name, 'prefix_string')) = 'dept2'  THEN 'V2
    ELSE (replace(column_name, 'prefix_string'))
    END
) = UPPER('v1');
```
In this case checking if it is just Constant Expression would be problematic.

maswin replied:
One problem we see with CASE expression is that it doesn't extract out the Column Domain while picking table layout. Switching it to AND/OR cases helps achieve it. Even if there is a column reference present in the equals side 
```
(CASE
    WHEN (replace(column_name, 'prefix_string')) = 'dept1' THEN 'V1'
    WHEN (replace(column_name, 'prefix_string')) = 'dept2'  THEN 'V2
    ELSE 'V3'
    END
) = column_name_2;
```
Above expression gets converted into a cumbersome AND/OR expression since column value's equality with THEN value cannot be evaluated and the expression won't be further simplified, but still it helps RowExpressionInterpretter/Optimizers in extracting column domain for the table from that expression. i.e, column_name_2 domain is set as [ V1, V2, V3].

kaikalur replied:
These things can lead to subtle bugs. And also if the other side is a function call pushing it into the CASE can make it evaluate many times. We should not need to do random stuff like upper(..). But maybe extend it to just constant or a variable/field reference/input expression.

maswin commented:
> Also please add some explicit end-to-end tests in hive connector to see the impact of this optimization.

Can you please point out where these tests exists? Internally we were able to see queries that ran for over 10 minutes to run in less than 10 seconds.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,187 @@
+                RowExpression operand = caseOperand.map(x -> logicalRowExpressions.equals(x, whenOperand)).orElse(whenOperand);
+                RowExpression result = ((SpecialFormExpression) whenClause).getArguments().get(1);
+                RowExpression resultExpression = isSwitchExpressionOnRight ?
+                        replaceArguments(comparison, valueExpression, result) : replaceArguments(comparison, result, valueExpression);
```
Add a helper method for this conditional so the logic is clear.

maswin replied:
Refactored the part.

kaikalur reviewed (changes requested):
We also need more end to end tests, not just unit test. Look at AbstractTestQueries - adding there will test with mutliple connectors which will be good.

kaikalur commented:
Also fix the test failures

maswin commented on presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java:
```diff
@@ -771,11 +771,29 @@ private static Form flip(Form binaryLogicalOperation)
+    }
+
+    public static RowExpression replaceArguments(CallExpression expression, RowExpression... arguments)
+    {
```
ExpressionUtils class was marked as Deprecated and mentioned that all RowExpression related logic to move to LogicalRowExpressions - 
https://github.com/prestodb/presto/commit/54c30164a43373d560f0219cc96f4e26c4c75af2

maswin commented on presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java:
```diff
@@ -771,11 +771,29 @@ private static Form flip(Form binaryLogicalOperation)
         return new CallExpression(argument.getSourceLocation(), "not", functionResolution.notFunction(), BOOLEAN, singletonList(argument));
     }
 
+    public RowExpression equals(RowExpression left, RowExpression right)
```
Done

maswin commented:
> We also need more end to end tests, not just unit test. Look at AbstractTestQueries - adding there will test with mutliple connectors which will be good.

Let me know if more cases needs to be handled in AbstractTestQueries

maswin commented:
I missed out a major case in the Optimizer.
```
(case 
    when condition1 then result1
    when condition2 then result2
    when condition3 then result3
else elseResult) = value
```
We currently convert the above expression to:
```
(result1 = value AND condition1)
OR
(result2 = value AND condition2)
OR
(result3 = value AND condition3)
OR
(elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
```
But we should actually be converting it to:
```
(result1 = value AND condition1)
OR
(result2 = value AND condition2 AND NOT(condition1))
OR
(result3 = value AND condition3 AND NOT(condition1) AND NOT(condition2))
OR
(elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
```
But this conversion makes the expression complicated and all conditions get re-evaluated multiple times. So I have added a check to determine if the conditions are all on same column and are mutually exclusive (TupleDomain does not overlap).
With the check we need not add the NOT condition since 
`condition1 AND NOT(condition2) => condition1 `
The domains are mutually exclusive and hence if condition1 is true it also means condition2 is false. 

Have also modified the AbstractTestQueries to assert the same query with and without the optimizer turned on and check if the results match.

kaikalur commented:
> I missed out a major case in the Optimizer.
> 
> ```
> (case 
>     when condition1 then result1
>     when condition2 then result2
>     when condition3 then result3
> else elseResult) = value
> ```
> 
> We currently convert the above expression to:
> 
> ```
> (result1 = value AND condition1)
> OR
> (result2 = value AND condition2)
> OR
> (result3 = value AND condition3)
> OR
> (elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
> ```
> 
> But we should actually be converting it to:
> 
> ```
> (result1 = value AND condition1)
> OR
> (result2 = value AND condition2 AND NOT(condition1))
> OR
> (result3 = value AND condition3 AND NOT(condition1) AND NOT(condition2))
> OR
> (elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
> ```
> 
> But this conversion makes the expression complicated and all conditions get re-evaluated multiple times. So I have added a check to determine if the conditions are all on same column and are mutually exclusive (TupleDomain does not overlap). With the check we need not add the NOT condition since `condition1 AND NOT(condition2) => condition1 ` The domains are mutually exclusive and hence if condition1 is true it also means condition2 is false.
> 
> Have also modified the AbstractTestQueries to assert the same query with and without the optimizer turned on and check if the results match.

That's why you should just restrict it to constants in the when clause as well as the result.

maswin commented:
> That's why you should just restrict it to constants in the when clause as well as the result.

As per your recommendation I have added check to see if final equals value is a constant/column reference so that we don't evaluate it multiple times. 
I am assuming you mean to restrict only cases where in "WHEN condition THEN result" the result needs to be constant. Not sure adding constant check to result will fix the above issue. Even if THEN returns a constant, the actual condition in WHEN clause needs to be negated unless the domains are mutually exclusive. for eg,
```
(CASE 
    WHEN col1=1 THEN 'a'
    WHEN col2=1 THEN 'b'
    ELSE 'c'
END) = 'b'
```
will be rewritten to:
```
('a'='b' AND col1=1) OR
('b'='b' AND col2=1) OR
('c'='b' AND col1<>1 AND col2<>2)
```
which will be simplified to
`col2=1`

But this is not a right simplification since if the rows have col1=1 it will fail the evaluation. So the right simplification should be: `col2 = 1 AND col1 <> 1`

Since we can't have the NOT check always I have implemented the domain to be mutually exclusive logic thereby avoiding the NOT condition.

[There seems to some flaky tests in Presto master branch, all the test failures I see in the PR currently seems to be in different part of code/tests]

kaikalur commented:
@maswin where are we at with this PR? This could be useful for one of our usecases so just wanted to check if there is anything else missing. Thanks!

maswin commented:
> @maswin where are we at with this PR? This could be useful for one of our usecases so just wanted to check if there is anything else missing. Thanks!

We have handled the case where there is a Cast done on top of the expression since most of the time to match the lhs and rhs expression type expressions are cast with a super data type. All other mentioned review comments are fixed and updated.

kaikalur commented:
Thank you! Added @rschlussel for further review and hopefully merging.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+        {
+            RowExpression commonLHSExpression = null;
+            List<RowExpression> rightHandSideExpressions = new ArrayList<>();
+            for (RowExpression whenClause : whenClauses) {
```
I find this a little tough to read (need to really work through the logic to understand what's going on)  I think it would be clearer to split up the three things going on in this loop even if it's less efficient.
e.g. something like
```

List<RowExpression> whenEqualityOperands = whenClauses.stream()
.map(whenClause -> (SpecialFormExpression) whenClause).getArguments().get(0))
.filter(logicalRowExpressions.isEqualExpression())
.collect(toImmutableList())

if(!whenEqualityOperands.size() == whenClauses.size()) {
   return false;
}

distinctLeftArgumentsCount = whenEqualityOperands.stream()
.map(whenOperand -> (CallExpression) whenOperand).getArguments().get(0))
.distinct()
.count()

if(!distinctLeftArgumentsCount == 1)
{
  return false;
}

List<RowExpression> rightArguments = whenEqualityOperands.stream()
.map(whenOperand -> ((CallExpression) whenOperand).getArguments().get(1))
.collect(toImmutableList())

return validateRHSExpressions(rightArguments);
```

maswin replied:
Broke the method into smaller ones.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+        private boolean canRewriteSearchedCaseExpression(List<RowExpression> whenClauses)
+        {
+            RowExpression commonLHSExpression = null;
+            List<RowExpression> rightHandSideExpressions = new ArrayList<>();
```
always use immutableCollections

maswin replied:
Done.

rschlussel reviewed (changes requested):
I added a bunch of comments to RewriteCaseExpressionPredicate, but I actually wonder if it would be better to do this with the other expression rewrites at the beginning of optimization using an ExpressionRewriteRuleSet

maswin commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,253 @@
+                .on(p -> p.filter(inputExpression, p.values(p.variable("col1"), p.variable("col2"), p.variable("col3"))))
+                .matches(filter(expectedRewrittenExpressionStr, values("col1", "col2", "col3")));
+
+        RowExpression expectedRewrittenExpression = testSqlToRowExpressionTranslator.translate(expectedRewrittenExpressionStr, TYPE_MAP);
```
The second half of the test checks the Optimized version of the expression. Most of the times the result=value condition gets evaluated to false and a simplified final expression would be produced. That simplification is not done by the rewriter and it is against unit testing principle to test it here, but it helped in understanding and developing the Rewriter better so retained them in the final commit.

maswin commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java:
```diff
@@ -165,6 +165,11 @@ protected void assertQuery(Session session, @Language("SQL") String actual, @Lan
         QueryAssertions.assertQuery(queryRunner, session, actual, expectedQueryRunner, expected, false, false, planAssertion);
     }
 
+    protected void assertQuery(Session actualSession, Session expectedSession, @Language("SQL") String sql)
```
Done.

maswin commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -5499,6 +5500,53 @@ public void testSwitchOptimization()
+        assertQuery(
+                caseExpressionRewriteEnabled,
+                caseExpressionRewriteDisabled,
+                "SELECT LINENUMBER FROM LINEITEM WHERE (CASE WHEN QUANTITY <= 15 THEN 'SMALL' WHEN (QUANTITY > 15 AND QUANTITY <= 30) THEN 'MEDIUM' ELSE 'LARGE' END) = 'SMALL'");
+
```
Didn't realize that. Fixed.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+                return null;
+            }
+
+            List<RowExpression> andExpressions = new ArrayList<>();
```
Done.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ * 2. The LHS expression in all WHEN clauses are the same.
+ *    For example, if one WHEN clause has a expression using col1 and another using col2, it will not work
+ * 3. The relational operator in the WHEN clause is equals. With other operators it is harder to check for exclusivity.
+ * 4. All the RHS expressions are a constant and unique
```
If we do the rewrite without this check (the RHS is not unique)
```
(case 
when col1=1 return 'a'
when col1=1 return 'b'
else 'c') = 'b'
```
will get simplified to
```
(col1=1 AND 'a'='b') OR
(col1=1 AND 'b'='b') OR
('c' = 'b' AND col1<>1 AND col2<>2)
```
`'a'='b'` condition would evaluate to false, and will get simplified to
`col1=1`
But this is a wrong simplification. If col1=1, it would have gone to first WHEN clause in CASE statement and returned 'a' which wouldn't be equal to 'b'. Our simplified condition allows rows with col1=1 to pass but in actual case it should not pass.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ * (result2 = value AND expression=constant2) OR
+ * (elseResult = value AND !(expression=constant1) AND !(expression=constant2))
+ * <p>
+ * This conversion happens when all of the following criteria are met:
```
Internally we saw many case statements with ~30-40 WHEN conditions and most of the time some WHEN return value gets matched to the final equals value and most of the OR clause gets removed in final expression, especially the last else statement. So I am not sure if this is a needed restriction.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ * <p>
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or column reference or input reference and not any function
+ * 2. The LHS expression in all WHEN clauses are the same.
```
I have rewritten the description with the reasoning for it.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ * (elseResult = value AND !(expression=constant1) AND !(expression=constant2))
+ * <p>
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or column reference or input reference and not any function
```
Will we be able to evaluate if the conditions are disjunct at query planning time if there are column references at RHS?

And most of the time it is the data visualization tool that generates such query, which usually is a column at LHS and constant at RHS. We have handled the most common case.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+            return searchedCase ? Optional.empty() : Optional.of(expression);
+        }
+
+        private boolean canRewriteSimpleCaseExpression(List<RowExpression> whenClauses)
```
RowExpression form of Simple and Searched were different. For SimpleCase expression WHEN clause was a ConstantExpression, for SearchedCase it was CallExpression(left: RowExpression, right: ConstantExpression, function: EQUALS). I had to do additional checks of whether the function is an EQUALS and extraction RHS expressions was different.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+            }
+        }
+
+        private RowExpression processCaseExpression(SpecialFormExpression caseExpression,
```
Most of the Case expression code parsing was copied from the following 2 classes:
https://github.com/prestodb/presto/blob/0571a805b3172af5e84abca9f5013e6a9b9842e4/presto-main/src/main/java/com/facebook/presto/sql/gen/SwitchCodeGenerator.java#L66
https://github.com/prestodb/presto/blob/0571a805b3172af5e84abca9f5013e6a9b9842e4/presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java#L628
So, Kind of assumed comments are not required. Do I still have to add comments?

rschlussel replied:
I think a comment would still be helpful so someone can understand this easily without having to jump into other classes to figure it out.

maswin replied:
Added.

maswin commented:
> I added a bunch of comments to RewriteCaseExpressionPredicate, but I actually wonder if it would be better to do this with the other expression rewrites at the beginning of optimization using an ExpressionRewriteRuleSet

We wanted the setting to be disabled initially, couldn't find disabling in ExpressionRewriteRuleSet/RowExpressionRewriteRuleSet. But yeah, that might help it extending to all types of Nodes.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+        {
+            Optional<RowExpression> caseOperand = getCaseOperand(caseExpression);
+            List<RowExpression> whenClauses;
+            RowExpression elseResult;
```
Done. Additionally made the check from ConstantExpression(NULL) = value to SpecialForm(IS_NULL, value).

maswin commented:
> > I added a bunch of comments to RewriteCaseExpressionPredicate, but I actually wonder if it would be better to do this with the other expression rewrites at the beginning of optimization using an ExpressionRewriteRuleSet
> 
> We wanted the setting to be disabled initially, couldn't find disabling in ExpressionRewriteRuleSet/RowExpressionRewriteRuleSet. But yeah, that might help it extending to all types of Nodes.

I have now made an additional commit that adds the ability for RowExpressionRewriterRuleSet to be disabled if required and modified the RewriteCaseExpressionPredicate to extend RowExpressionRewriteRuleSet. This enables it to optimize filter expression in JOIN condition too.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ *      RowExpressionDomainTranslator can construct a Domain for the column
+ */
+public class RewriteCaseExpressionPredicate
+        implements Rule<FilterNode>
```
Fixed.

maswin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,329 @@
+                        value);
+            }
+            else {
+                return processCaseExpression(
```
Done.

rschlussel reviewed (approved):
Thank you!

rschlussel commented:
@kaikalur can you review/approve?  Merging is blocked since you are still marked as requesting changes.

rschlussel commented:
yup.  I'll merge now!
